Refactoring of app.py into modules

Colex opened the pull request:
https://assembly.com/formspree/bounties/17

I have reduced app.py into a few lines by splitting the code (mostly) into views.py and models.py.
The model contains all the business logic whilst views.py takes care of responding to the requests properly.
The project is not completely refactored, I have tried to change the least code possible while moving the code around (the logic is still exactly the same).
I'll try to add more comments and maybe a few changes soon. In the meantime, feel free to check it and change it.

colevscode commented on forms/app.py:
```diff
@@ -1,379 +1,15 @@
-    app.config.from_object(settings)
-
-    DB.init_app(app)
-    configure_routes(app)
```
Why remove the create_app() function? Is there some advantage to having a globally scoped app variable?

Colex replied:
I was only using app to define routing, so no advantages as I can revert it and do it in other ways. If you prefer routing without decorators, I can also revert that.

colevscode replied:
Ah got it. No you're right, I tend to see decorators used more commonly in
flask apps that the way we were doing it.

On Fri, Jan 9, 2015 at 9:23 PM, Alexandre Santos <notifications@github.com>
wrote:

> In forms/app.py
> <https://github.com/asm-products/formspree/pull/11#discussion-diff-22755332>
> :
>
> > -
> > -def configure_routes(app):
> > -    app.add_url_rule('/', 'index', view_func=default, methods=['GET'])
> > -    app.add_url_rule('/favicon.ico', view_func=favicon)
> > -    app.add_url_rule('/<email>', 'send', view_func=send, methods=['GET', 'POST'])
> > -    app.add_url_rule('/confirm/<nonce>', 'confirm_email', view_func=confirm_email, methods=['GET'])
> > -    app.add_url_rule('/thanks', 'thanks', view_func=thanks, methods=['GET'])
> > -    app.add_url_rule('/<path:template>', 'default', view_func=default, methods=['GET'])
> > -
> > -
> > -def create_app():
> > -    app = flask.Flask(__name__)
> > -    app.config.from_object(settings)
> > -
> > -    DB.init_app(app)
> > -    configure_routes(app)
>
> I was only using app to define routing, so no advantages as I can revert
> it and do it in other ways. If you prefer routing without decorators, I can
> also revert that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/asm-products/formspree/pull/11/files#r22755332>.
>

fiatjaf commented on forms/consts.py:
```diff
@@ -0,0 +1,7 @@
+
+HASH = lambda x, y: hashlib.md5(x+y+settings.NONCE_SECRET).hexdigest()
+IS_VALID_EMAIL = lambda x: re.match(r"[^@]+@[^@]+\.[^@]+", x)
+EXCLUDE_KEYS = ['_gotcha', '_next', '_subject', '_cc']
```
I think this could be merged into helpers.py (but that's only because I get more confused the more files a project has), what do you think?

Colex replied:
Yes, I agree. I would also suggest making IS_VALID_EMAIL and HASH named functions within helpers .py instead of "constants".

colevscode commented:
Awesome thanks Alexandre.

Sorry for not being more specific, but when I said to move forms into it's
own module, I mean a separate "forms" folder with an _init_.py, models.py
and views.py, just for form related operations.

I realize that having the forms toplevel folder with a forms nested folder
is a bit confusing. Maybe the toplevel folder should be called formspree.

Re consts/ utils, I agree with merging them.

On Fri, Jan 9, 2015 at 8:28 PM, Alexandre Santos <notifications@github.com>
wrote:

> https://assembly.com/formspree/bounties/17
>
> I have reduced app.py into a few lines by splitting the code (mostly) into
> views.py and models.py.
> The model contains all the business logic whilst views.py takes care of
> responding to the requests properly.
> The project is not completely refactored, I have tried to change the least
> code possible while moving the code around (the logic is still exactly the
> same).
> I'll try to add more comments and maybe a few changes soon. In the
> meantime, feel free to check it and change it.
> ------------------------------
> You can merge this Pull Request by running
>
>   git pull https://github.com/Colex/formspree refactoring
>
> Or view, comment on, or merge it at:
>
>   https://github.com/asm-products/formspree/pull/11
> Commit Summary
>
>    - Split app.py into models, views and helpers
>    - Merge pull request #10 from fiatjaf/migrate
>
> File Changes
>
>    - *M* forms/__init__.py
>    <https://github.com/asm-products/formspree/pull/11/files#diff-0> (5)
>    - *M* forms/app.py
>    <https://github.com/asm-products/formspree/pull/11/files#diff-1> (376)
>    - *A* forms/consts.py
>    <https://github.com/asm-products/formspree/pull/11/files#diff-2> (7)
>    - *A* forms/helpers.py
>    <https://github.com/asm-products/formspree/pull/11/files#diff-3> (101)
>    - *A* forms/models.py
>    <https://github.com/asm-products/formspree/pull/11/files#diff-4> (122)
>    - *M* forms/settings.py
>    <https://github.com/asm-products/formspree/pull/11/files#diff-5> (4)
>    - *A* forms/views.py
>    <https://github.com/asm-products/formspree/pull/11/files#diff-6> (133)
>    - *M* manage.py
>    <https://github.com/asm-products/formspree/pull/11/files#diff-7> (12)
>
> Patch Links:
>
>    - https://github.com/asm-products/formspree/pull/11.patch
>    - https://github.com/asm-products/formspree/pull/11.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/asm-products/formspree/pull/11>.
>

Colex commented:
No worries, Cole. Thank you for the feedback.
I moved things around a little bit and reverted a few changes (routing with decorators mainly, some people might not like it, and it's easy to go to a routes.py file and find out exactly what each route is doing, so that's what I implemented for now). Splitting the business logic from presentation remains the same. I tried to make it so developers don't need to make changes to startup scripts (app.py or __init__.py) and may focus on the features. 
I also thought of easily supporting different environments on settings, but I would like to reduce the number of changes for now.
I can iterate over it again if there might be things you would like to see different. (Also, let me know if you're looking for something completely different).

colevscode commented:
Woops, sorry, I meant merging helpers and consts. I think there should
remain a separate top-level utils.py with app-wide utilities in it. I think
there can still be a forms-specific helpers.py with those "constants" and
referrer_to_path, http_form_to_dict, and send_email functions.

Other than that this looks good.

The only other feedback I have is that this merge doesn't seem to run.
Please get it working on localhost so we can test before merging.

On Sat, Jan 10, 2015 at 9:01 AM, Alexandre Santos <notifications@github.com>
wrote:

> No worries, Cole. Thank you for the feedback.
> I moved things around a little bit and reverted a few changes (routing
> with decorators mainly, some people might not like it, and it's easy to go
> to a routes.py file and find out exactly what each route is doing, so
> that's what I implemented for now). Splitting the business logic from
> presentation remains the same. I tried to make it so developers don't need
> to make changes to startup scripts (app.py or *init*.py) and may focus on
> the features.
> I also thought of easily supporting different environments on settings,
> but I would like to reduce the number of changes for now.
> I can iterate over it again if there might be things you would like to see
> different. (Also, let me know if you're looking for something completely
> different).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/asm-products/formspree/pull/11#issuecomment-69456750>.
>

Colex commented:
Hey Cole, it does make sense to have a shared utils app-wide, I just took your comment too literally. Nevertheless, I reverted that and merged helpers and "consts".
I understand that the project doesn't have tests yet, so I tried testing all functionalities manually (send confirmation, confirm, send emails). Everything I tested worked fine.
